core/src/main/scala/org/apache/spark/SparkContext.scala

```diff
@@ -1093,7 +1093,7 @@ class SparkContext(config: SparkConf) extends Logging with ExecutorAllocationCli
   def addFile(path: String, recursive: Boolean): Unit = {
     val uri = new URI(path)
     val schemeCorrectedPath = uri.getScheme match {
-      case null | "local" => "file:" + uri.getPath
+      case null | "local" => "file:" + new File(path).getCanonicalPath
```

Comment:
I think this might be slightly more robust as `new File(path).getCanonicalFile.toURI`. Then the `"file:"` can go away too.